src/crimson/thread/ThreadPool.cc

```diff
@@ -0,0 +1,77 @@
+#include <chrono>
+
+#include "ThreadPool.h"
+#include "crimson/net/Config.h"
+
+namespace ceph::thread {
+
+ThreadPool::ThreadPool(size_t n)
+  : threads{create_threads(n)}
+{
+  // try to shard the queue evenly
+  const auto queue_size = num_queue_size / seastar::smp::count;
+  for (size_t i = 0; i < seastar::smp::count; i++) {
+    if (i == 0) {
+      free_slots.emplace_back(queue_size + num_queue_size % queue_size);
+    } else {
+      free_slots.emplace_back(queue_size);
+    }
+  }
+}
+
+ThreadPool::~ThreadPool()
+{
+  stop = true;
+  cond.notify_all();
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+std::vector<std::thread> ThreadPool::create_threads(size_t n)
+{
+  std::vector<std::thread> workers;
+  for (size_t i = 0; i < n; i++) {
+    workers.emplace_back([this] {
```

Comment:
we'll eventually want a way to prevent these threads from running on cores with a seastar reactor. that can come later though?

Reply:
we could. but i assume that these threads will not as busy as the reactor threads. in most of their time, they are blocked by blocking calls. and it would be better to have `seastar::smp::count` reactors to take the full advantage of cores in the system. also, the reactor thread will try to be nice using `pause` instruction to yield the CPU when it has nothing to do. so i guess the alien threads won't be starved.